plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -0,0 +1,349 @@
+arbitrarily, otherwise we'll be unable to read back old transactions and validate them.
+
+Consequently we don't use the derivable instances of `Flat` for the PLC types that go
+on the chain.
+
+However, the instances in this file *are* constrained by instances for names, type names,
+and annotations. What's to stop the instances for *those* changing, thus changing
+the overall encoding on the chain?
+
+The answer is that what goes on the chain is *always* a `Program TyName Name ()`. The instances
+for `TyName` and `Name` are nailed down here, and the instance for `()` is standard.
+
+However, having this flexibility allows us to encode e.g. PLC with substantial annotations
+(like position information) in situation where the stability is *not* critical, such as
+for testing.
+-}
+
+{- Note [Encoding/decoding constructor tags using Flat]
+Flat saves space when compared to CBOR by allowing tags to use the minimum
+number of bits required for their encoding.
+
+This requires specialised encode/decode functions for each constructor
+that encodes a different number of possibilities. Here is a list of the
+tags and their used/available encoding possibilities.
+
+| Data type       | Function          | Used | Available |
+|-----------------|-------------------|------|-----------|
+| Static builtins | encodeBuiltin     | 22   | 32        |
+| Kinds           | encodeKind        | 2    | 2         |
+| Types           | encodeType        | 7    | 8         |
+| BuiltinNames    | encodeBuiltinName | 2    | 2         |
+| Terms           | encodeTerm        | 10   | 16        |
+
+For format stability we are manually assigning the tag values to the
+constructors (and we do not use a generic algorithm that may change this order).
+
+Flat uses an extra function in its class definition called `size`. Since we want
+to reserve some space for future data constructors and we don't want to have the
+sizes desynchronised from the encoding and decoding functions we have manual
+implementations for them (if they have any constructors reserved for future use).
+-}
+
+{- Note [Don't use catamorphims!]
+We use Codec.Serialise for encoding.  This uses an intermediate type
+`Encoding` to encode things. `Encoding` is a monoid, which allows
+subobjects to be encoded and then efficiently concatenated when a
+larger object is being encoded to Flat.  The monoid structure makes it
+tempting to use catamorphisms to encode things, but this is
+*dangerous*. When a list is encoded to Flat, the start of the list is
+marked by a special token; this is then followed by the encodings of
+the individual list elements, then another token to mark the end of
+the list.  If we use a catamorphism to encode a list `l` it just
+encodes the elements and concatenates the encoded versions together
+without the markers, which is wrong: instead we should call `encode l`
+directly.
+-}
+
+-- See Note [The G, the Tag and the Auto].
```

Comment:
Is this a reference you want?

Reply:
It is the same case as it is for CBOR, which is the case for most other comments.